### ISSUE TITLE
kube-dns never resolves if a domain returns NOERROR with 0 answer records once

### ISSUE THREAD

**Issue by ahmetb** [kubernetes/dns#121]
tl;dr If a nameserver replies status=NOERROR with no answer section to a DNS A question, kube-dns always caches this result. If the domain name actually gets an A record _after_ it's queried through kube-dns, it never (I waited a few days) resolves from the pods, but does resolve outside the container (e.g. on my laptop) just fine.

#### Prerequisites

- Have a domain name `alp.im` and the nameservers are pointed to CloudFlare.
- Have nslookup/dig installed on your workstation.
- Have a minikube cluster ready on your workstation
    - running kubernetes v1.6.0
    - kube-dns comes by default, running `gcr.io/google_containers/k8s-dns-kube-dns-amd64:1.14.1`

#### Step 1: Domain does not exist, query from your laptop

Note `ANSWER: 0`, and `status: NOERROR`

```
$ dig A z.alp.im

; <<>> DiG 9.8.3-P1 <<>> A z.alp.im
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 64978
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;z.alp.im.			IN	A

;; AUTHORITY SECTION:
alp.im.			1799	IN	SOA	ivan.ns.cloudflare.com. dns.cloudflare.com. 2025042470 10000 2400 604800 3600

;; Query time: 196 msec
;; SERVER: 2401:fa00:fa::1#53(2401:fa00:fa::1)
;; WHEN: Thu Jun 29 10:51:35 2017
;; MSG SIZE  rcvd: 99
```

#### Step 2: Domain does not exist, query from Pod on Kubernetes

Start a toolbelt/dig container with shell and run the same query:

> ⚠️ Do not exit this container as you will reuse it later. 

Note the response is the same, `ANSWER: 0` and `NOERROR`.

```
$ kubectl run -i -t --rm --image=toolbelt/dig dig --command -- sh
If you don't see a command prompt, try pressing enter.
/ # dig A z.alp.im

; <<>> DiG 9.11.1-P1 <<>> A z.alp.im
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 11209
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;z.alp.im.			IN	A

;; AUTHORITY SECTION:
alp.im.			1724	IN	SOA	ivan.ns.cloudflare.com. dns.cloudflare.com. 2025042470 10000 2400 604800 3600

;; Query time: 74 msec
;; SERVER: 10.0.0.10#53(10.0.0.10)
;; WHEN: Thu Jun 29 17:55:46 UTC 2017
;; MSG SIZE  rcvd: 99
```

(Also note that `SERVER: 10.0.0.10#53` which is kube-dns.)

#### Step 3: Create an A record for the domain

Here I use CloudFlare as it manages my DNS. 

![image](https://user-images.githubusercontent.com/159209/27702492-16f4ce86-5cb9-11e7-9cf3-bda59ef18660.png)

#### Step 4: Test DNS record from your laptop

Run `dig` on your laptop (note `;; ANSWER SECTION:` and `8.8.8.8` answer):

```
$ dig A z.alp.im

; <<>> DiG 9.8.3-P1 <<>> A z.alp.im
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 37570
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;z.alp.im.			IN	A

;; ANSWER SECTION:
z.alp.im.		299	IN	A	8.8.8.8

;; Query time: 196 msec
;; SERVER: 2401:fa00:fa::1#53(2401:fa00:fa::1)
;; WHEN: Thu Jun 29 10:54:44 2017
;; MSG SIZE  rcvd: 53
```

#### Step 5: Test DNS record from Pod on Kubernetes

Run the same command again:

```
/ # dig A z.alp.im

; <<>> DiG 9.11.1-P1 <<>> A z.alp.im
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 45420
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;z.alp.im.			IN	A

;; Query time: 0 msec
;; SERVER: 10.0.0.10#53(10.0.0.10)
;; WHEN: Thu Jun 29 18:00:24 UTC 2017
;; MSG SIZE  rcvd: 37
```

Note the [diff](https://www.diffchecker.com/l2GTkuBd):

- still `ANSWER: 0` and `status: NOERROR` (but it resolves just fine outside the cluster)
- `;; AUTHORITY SECTION:` disappeared and `AUTHORITY:` changed to `0` from the previous time we ran this. 
- `;; Query time: 0 msec` (was 79 ms) –I assume it means it's just a cached response.
    - Query time stays as 0 ms no matter how many times I run the same command.

## What else I tried

- **Try it on GKE**: I tried with k8s v1.5.x and v1.6.4.  &rarr; Same issue. (cc: @bowei)
- **Query from a different pod** on minikube: I started a new Pod and queried from there  &rarr; Same issue.
- **Restart kube-dns Pod** &rarr; This **worked** on GKE, but **not** on minikube.

      $ kubectl delete pods -n kube-system -l k8s-app=kube-dns
      pod "kube-dns-268032401-69xk5" deleted

## Impact

I am not sure why this has not been discovered before. I noticed this behavior while using kube-lego on GKE. Once kube-lego applies for a TLS certificate, it polls the domain name of the service (e.g. `example.com/.well-known/<token>`) before asking Let's Encrypt to validate it. Before I create an Ingress with kube-lego annotation, I don't have the external IP yet so I can't configure the domain, but the kube-lego Pod already picks it up and starts querying my domain in an infinite loop. It never succeeds because first time it looked up the hostname, the A record didn't exist, so that result is cached forever. After I add A record, it still can't resolve. The moment I delete kube-dns Pods and they get recreated, it immediately starts working and resolves the hostname and completes the kube-lego challenge.

**Comment by bowei** [comment]
@ahmetb I think it's legal in DNS to cache if we get a rcode == 0 response with 0 entries. This looks to be the behavior of the Cloudflare server (sending rcode 0 instead of NXDOMAIN). It looks like the TTL was around 30 minutes for the DNS record. If the record is going to be changing, it would be advisable to reduce the TTL to get faster cache updates.

**Comment by ahmetb** [comment]
@bowei I think @viglesiasce reproduced this with Google Domains (or Cloud DNS) too.

In my experience, the cache was not updated even after 24 hours when I left it at that.

**Comment by jpap** [comment]
Linking to #119 with respect to Cloudflare.

**Comment by vavrusa** [comment]
RCODE=0 with no response is the NODATA pseudo-rcode. For the purpose of caching, it shouldn't be treated differently from NXDOMAIN with one exception - it doesn't say anything about non-existence of names below the requested name. See https://tools.ietf.org/html/rfc2308#section-2.2 for guideline. It's possibly related to https://github.com/miekg/dns/issues/428

**Comment by ahmetb** [comment]
I reported this to various folks at CloudFlare, still waiting a response. However, if anyone can help pinpoint where the caching happens, under what circumstances and why it lasts so long (i.e. >24h or in my experience, indefinitely), those would help fixing this problem, too.

**Comment by vavrusa** [comment]
I work at Cloudflare, so happy to answer any questions. It's not however specific to Cloudflare DNS; NODATA is a kind of answer you get from an authoritative server when the requested name exists, but the record type you're looking for doesn't, which is quite common. The RFC2038 I linked provides a guideline on how clients should implement negative caching for all cases of negative answers - hope that helps.

**Comment by ahmetb** [comment]
Got an answer from the CloudFlare support:

> We are aware of this behavior and it has been escalated previously to our DNS team. Their response is that at this moment we could not change/improve this behaviour. It's our design feature.
>
> We are aware of this and will be working in improving the behavior in the future - but this will not happen earlier than 6 months.
>
> Unfortunately, at this point, this is how our DNS is working.

We should look at fixing the caching behavior in kube-dns (or miekg/dns, or wherever it is) as a mitigation. Not caching 0-record answers sounds like it would yield a low-impact cache-miss rate to me. @bowei thoughts?

**Comment by bowei** [comment]
@ahmetb caching is done with dnsmasq (http://www.thekelleys.org.uk/dnsmasq/doc.html) with no special tuneables. Maybe there is a flag that can disable caching that response? I'm surprised this does not impact more people, not just users of Kubernetes. dnsmasq is a popular piece of software, standard resolver on some Linux distros.

**Comment by hugorut** [comment]
I have this exact same issue (using kube-lego on GCE) but using google cloud DNS.  Kube lego cannot resolve my domain when trying to request the token in order to issue a certificate. External to any kube pod the domain name resolves fine. When digging the domain within the pod it still gets `ANSWER: 0` and `status: NOERROR`. 

I tried restarting the kube-dns with 
```
kubectl delete pods -n kube-system -l k8s-app=kube-dns
```
but to no avail.

Is there anything I can do to expedite invalidating the DNS cache or is it a matter of waiting it out?

**Comment by bowei** [comment]
Can you post the output of `dig` for the entry? dnsmasq uses the TTL of the SOA record for negative replies, otherwise it will be 0.

**Comment by hugorut** [comment]
From the pod:

```
; <<>> DiG 9.10.4-P8 <<>> my.domain.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NXDOMAIN, id: 34318
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;api.striing.io.			IN	A

;; Query time: 0 msec
;; SERVER: 10.39.240.10#53(10.39.240.10)
;; WHEN: Wed Aug 02 06:45:50 UTC 2017
;; MSG SIZE  rcvd: 43
```

but i will occasionally get this answer instead:

```
; <<>> DiG 9.10.4-P8 <<>> my.domain.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NXDOMAIN, id: 18668
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;api.striing.io.			IN	A

;; AUTHORITY SECTION:
my.domain.com		895	IN	SOA	ns-1660.awsdns-15.co.uk. awsdns-hostmaster.amazon.com. 1 7200 900 1209600 86400

;; Query time: 1 msec
;; SERVER: 10.39.240.10#53(10.39.240.10)
;; WHEN: Wed Aug 02 06:35:03 UTC 2017
;; MSG SIZE  rcvd: 130
```

This SOA is from my AWS which was my prior DNS provider.

**Comment by bowei** [comment]
I will try playing with dnsmasq flags and see if we can change its negative caching behavior.

**Comment by ahmetb** [comment]
@bowei Any luck?

**Comment by coresolve** [comment]
Looks like ` - --no-negcache` added to the dnsmasq args ought to do it. 

Credit to https://rsmitty.github.io/KubeDNS-Tweaks/

**Comment by ahmetb** [comment]
@coresolve whoa this is amazing. @bowei do you think it's sensible to incorporate this as a default in kube-dns distribution?

**Comment by bowei** [comment]
yes, since we don't enable neg caching

**Comment by cblecker** [comment]
This would be just in https://github.com/kubernetes/kubernetes/blob/e557f184e2bf563c9d50719ca0942cc6128909d7/cluster/addons/dns/kube-dns.yaml.base#L158-L169, right?

**Comment by bowei** [comment]
yes, that should be a one-line change to the yaml

**Comment by cblecker** [comment]
Opened https://github.com/kubernetes/kubernetes/pull/53604 to add this

**Comment by miekg** [comment]
Has anyone looked into the impact of removing negative caching on the volume of DNS requests that now need resolved again and again?

**Comment by ahmetb** [comment]
@miekg I think we don't know what will this change break. However, unless changed, many software that rely on domains eventually resolving stays broken. I'm not sure if we have enough tools to answer this question properly.

**Comment by krogon** [comment]
Why did we disable neg-caching as default instead of setting reasonable TTL value with `--neg-ttl=600`?
With huge amount of queries in kubernetes related to ndots settings this would have negative impact.
